File: Coursera/Algorithms_Part_2/Ex38KnuthMorrisStringSearch.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdOut;

public class Ex38KnuthMorrisStringSearch {

    private static final int R = 256;
    private final String pattern;
    private final int[][] pfa;
    private final int m;

    public Ex38KnuthMorrisStringSearch(String pattern) {
        this.pattern = pattern;
        this.m = pattern.length();

        this.pfa = new int[R][m];
        this.pfa[pattern.charAt(0)][0] = 1;
        int x = 0;
        for (int i = 1; i < m; i++) {
            for (int j = 0; j < R; j++)
                this.pfa[j][i] = this.pfa[j][x];
            this.pfa[pattern.charAt(i)][i] = i + 1;
            x = this.pfa[pattern.charAt(i)][x];
        }
    }

    public int search(String text) {
        for (int i = 0, j = 0; i < text.length(); i++) {
            j = pfa[text.charAt(i)][j];
            if (j == m) return i - m + 1;
        }
        return -1;
        // return search(new In(text));
    }

    public int search(In in) {
        for (int i = 0, j = 0; !in.isEmpty(); i++) {
            j = pfa[in.readChar()][j];
            if (j == m) return i - m + 1;
        }
        return -1;
    }

    public String pattern() {
        return pattern;
    }

    private static void test(String pattern, String text) {
        Ex38KnuthMorrisStringSearch kmSearch = new Ex38KnuthMorrisStringSearch(pattern);
        StringBuilder builder = new StringBuilder();
        int index = kmSearch.search(text);
        if (index > -1) {
            builder.append(text);
            builder.insert(index + pattern.length(), '*');
            builder.insert(index, '*');
        }
        else builder.append("not found");
        StdOut.println("search for " + pattern + " in " + text + ": " + builder.toString());
    }

    public static void main(String[] args) {
        test("ABRA", "ABACADABRAC");
        test("BAAAAAAAAA", "ABAAAABAAAAAAAAA");
        test("BAAAABAAAA", "ABAAAABAAAAAAAAA");
        test("BAAABAAAA", "ABAAAABAAAAAAAAA");
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex38KnuthMorrisStringSearch.py
def search(pattern, text):
    n = len(text)
    m = len(pattern)
    R = 256

    pfa = [[0] * R] * m
    x = 0
    for i in range(m):
        x = pfa[x][ord(pattern[i])]
        pfa[i] = pfa[x].copy()
        pfa[i][ord(pattern[i])] = i + 1

    j = 0
    for i in range(n):
        j = pfa[j][ord(text[i])]
        if j == m:
            return i - j + 1

    return -1


def test_knuth_morris_string_search():

    def test(pattern, text):
        index = search(pattern, text)
        if index != -1:
            builder = text[:index] + '*' + \
                text[index:index+len(pattern)] + '*' + \
                text[index+len(pattern):]
        else:
            builder = 'not found'
        print(f'search for "{pattern}" in "{text}": {builder}')

    test("ABRA", "ABACADABRAC")
    test("BAAAAAAAAA", "ABAAAABAAAAAAAAA")
    test("BAAAABAAAA", "ABAAAABAAAAAAAAA")
    test("BAAABAAAA", "ABAAAABAAAAAAAAA")


if __name__ == '__main__':
    test_knuth_morris_string_search()
